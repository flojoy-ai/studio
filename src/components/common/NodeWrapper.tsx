import React from "react";
import { CustomNodeProps } from "../../types/node";

const NodeWrapper = ({
  wrapperProps: { wrapperOnClick, nodeError },
  children,
}: {
  wrapperProps: CustomNodeProps;
  children: React.ReactNode;
}) => {
  return (
    <div
      className="relative"
      data-testid="node-wrapper"
      onClick={wrapperOnClick}
    >
      {nodeError && <ErrorPopup message={nodeError} />}
      {children}
    </div>
  );
};

export default NodeWrapper;

const ErrorPopup = ({ message }: { message: string }) => {
  return (
    <div
      className="absolute -top-14 left-1/2 right-1/2 z-50 my-3 w-max max-w-md -translate-x-1/2 -translate-y-1/2 rounded-md border-2 border-solid border-red-400 p-3 font-semibold"
      data-testid="node-error-popup"
    >
      <p className="m-0 font-sans text-lg text-gray-900 dark:text-gray-50">
        {message}
      </p>
      <div className="absolute -bottom-3 left-1/2 -ml-2 h-0 w-0 border-x-8 border-t-8 border-solid border-x-transparent border-t-red-400" />
    </div>
  );
};
