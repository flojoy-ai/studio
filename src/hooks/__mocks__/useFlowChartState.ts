export const useFlowChartState = jest.fn(() => ({
  rfInstance: {
    nodes: [],
  },
  nodes: [],
  edges: [],
  updateCtrlInputDataForNode: jest.fn(),
  removeCtrlInputDataForNode: jest.fn(),
  ctrlsManifest: [
    {
      type: "input",
      name: "Slider",
      id: "INPUT_PLACEHOLDER",
      hidden: false,
      minHeight: 1,
      minWidth: 2,
    },
  ],
  setCtrlsManifest: jest.fn(),
  isEditMode: false,
  gridLayout: [],
  setGridLayout: jest.fn(),
  setRfInstance: jest.fn(),
  loadFlowExportObject: jest.fn(),
  openFileSelector: jest.fn(),
  saveFile: jest.fn(),
  setIsEditMode: jest.fn(),
  uiTheme: "dark",
  setUiTheme: jest.fn(),
  showLogs: false,
  setShowLogs: jest.fn(),
  runningNode: "",
  setRunningNode: jest.fn(),
  failedNode: "",
  setFailedNode: jest.fn(),
  setEdges: jest.fn(),
  setNodes: jest.fn(),
}));
