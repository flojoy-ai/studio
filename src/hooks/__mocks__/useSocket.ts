import { IServerStatus } from "@src/context/socket.context";
module.exports = {
  useSocket: jest.fn(() => ({
    states: {
      socketId: "socket-1234",
      serverStatus: IServerStatus.CONNECTING,
      setProgramResults: jest.fn(),
      failureReason: 'test error',
      failedNode: 'node-123'
    },
  })),
};
