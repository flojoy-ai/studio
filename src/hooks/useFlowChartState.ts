import { ElementsData } from "@/types";
import { atom, useAtom } from "jotai";
import { atomWithImmer } from "jotai-immer";
import localforage from "localforage";
import { ReactFlowJsonObject } from "reactflow";

export type CtrlManifestParam = ElementsData["ctrls"][""] & {
  nodeId: string;
  id: string;
};
export interface PlotManifestParam {
  node: string;
  plot?: PlotType;
  input?: string;
  output?: string;
}

interface PlotType {
  type: string;
  mode: string;
}

export interface CtlManifestType {
  type: string;
  name: string;
  id: string;
  param?: PlotManifestParam | CtrlManifestParam | string;
  val?: string | number | boolean;
  hidden?: boolean;
  segmentColor?: string;
  controlGroup?: string;
  label?: string;
  minHeight: number;
  minWidth: number;
  layout: ReactGridLayout.Layout;
}
export interface EnvVarCredentialType {
  key: string;
  value: string;
}

export const failedNodeAtom = atom<string>("");
export const runningNodeAtom = atom<string>("");
export const nodeStatusAtom = atom((get) => ({
  runningNode: get(runningNodeAtom),
  failedNode: get(failedNodeAtom),
}));

const showLogsAtom = atomWithImmer<boolean>(false);
const rfInstanceAtom = atomWithImmer<
  ReactFlowJsonObject<ElementsData> | undefined
>(undefined);
const editModeAtom = atomWithImmer<boolean>(false);
const credentialsAtom = atomWithImmer<EnvVarCredentialType[]>([]);
const isSidebarOpenAtom = atom<boolean>(false);
const nodeParamChangedAtom = atom<boolean>(false);
export const centerPositionAtom = atom<{ x: number; y: number } | undefined>(
  undefined,
);
localforage.config({ name: "react-flow", storeName: "flows" });

export function useFlowChartState() {
  const [rfInstance, setRfInstance] = useAtom(rfInstanceAtom);
  const [isEditMode, setIsEditMode] = useAtom(editModeAtom);
  const [showLogs, setShowLogs] = useAtom(showLogsAtom);
  const [runningNode, setRunningNode] = useAtom(runningNodeAtom);
  const [failedNode, setFailedNode] = useAtom(failedNodeAtom);
  const [credentials, setCredentials] = useAtom(credentialsAtom);
  const [isSidebarOpen, setIsSidebarOpen] = useAtom(isSidebarOpenAtom);
  const [nodeParamChanged, setNodeParamChanged] = useAtom(nodeParamChangedAtom);

  return {
    rfInstance,
    setRfInstance,
    isEditMode,
    setIsEditMode,
    showLogs,
    setShowLogs,
    runningNode,
    setRunningNode,
    failedNode,
    setFailedNode,
    credentials,
    setCredentials,
    nodeParamChanged,
    setNodeParamChanged,
    isSidebarOpen,
    setIsSidebarOpen,
  };
}
