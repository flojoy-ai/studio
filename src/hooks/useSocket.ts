import { useContext } from "react";
import { SocketContext } from "../context/socket.context";

export const useSocket = () => {
  const data = useContext(SocketContext);
  if (!data) {
    throw new Error("useSocket must be used within a SocketProvider");
  }
  return { ...data };
};
