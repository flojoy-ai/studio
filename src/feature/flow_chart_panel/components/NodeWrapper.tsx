import { Box, Text, createStyles } from "@mantine/core";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import { useSocket } from "@src/hooks/useSocket";
import React, { useEffect, useState } from "react";
import { CustomNodeProps } from "../types/CustomNodeProps";
import NodeButtons from "./node-edit-menu/NodeButtons";

const NodeWrapper = ({
  data,
  children,
}: CustomNodeProps & {
  children: React.ReactNode;
}) => {
  const { setIsEditMode, setIsExpandMode } = useFlowChartState();
  const { failedNode } = useFlowChartState();
  const { states } = useSocket();
  const [runError, setRunError] = useState<{
    message: string;
    show: boolean;
  } | null>(null);

  useEffect(() => {
    console.log("15");
    if (failedNode === data.id) {
      setRunError({
        message: states?.failureReason ?? "Unknown failure reason",
        show: false,
      });
    }
    return () => {
      setRunError(null);
    };
  }, [failedNode, states?.failureReason]);

  return (
    <Box
      data-testid="node-wrapper"
      pos="relative"
      onClick={() => setIsEditMode(true)}
    >
      {data.selected && (
        <NodeButtons data={data} setIsExpandMode={setIsExpandMode} />
      )}
      {runError && <ErrorPopup message={runError.message} />}
      {children}
    </Box>
  );
};

export default NodeWrapper;

const useStyles = createStyles((theme) => {
  return {
    popupContainer: {
      position: "absolute",
      top: -50,
      left: "50%",
      right: "50%",
      transform: "translate(-50%, -50%)",
      backgroundColor: "#dcf8c6",
      borderRadius: theme.radius.sm,
      margin: "10px 0",
      padding: 10,
      width: "max-content",
      maxWidth: 500,
      zIndex: 100,
      color: theme.black,
      fontWeight: 600,
    },
    popupArrow: {
      borderStyle: "solid",
      borderWidth: "10px 10px 0 10px",
      borderColor: "#dcf8c6 transparent transparent transparent",
      height: 0,
      width: 0,
      position: "absolute",
      bottom: -10,
      left: "50%",
      marginLeft: -10,
    },
  };
});

const ErrorPopup = ({ message }: { message: string }) => {
  const { classes } = useStyles();
  return (
    <Box className={classes.popupContainer} data-testid="node-error-popup">
      <Text fz="lg">{message}</Text>
      <Box className={classes.popupArrow} />
    </Box>
  );
};
