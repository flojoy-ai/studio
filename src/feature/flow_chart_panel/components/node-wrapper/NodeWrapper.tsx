import React, { useEffect, useState } from "react";
import { CustomNodeProps } from "../../types/CustomNodeProps";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import HandleComponent from "../HandleComponent";
import "./node-wrapper.css";
import { useSocket } from "@src/hooks/useSocket";

const NodeWrapper = ({
  data,
  children,
}: CustomNodeProps & {
  children: React.ReactNode;
}) => {
  const { failedNode } = useFlowChartState();
  const { states } = useSocket();
  const [runError, setRunError] = useState<{
    message: string;
    show: boolean;
  } | null>(null);

  const handleMouseEnter = () => {
    if (runError) {
      setRunError((p) => ({ ...p!, show: true }));
    }
  };
  const handleMouseLeave = () => {
    if (runError) {
      setRunError((p) => ({ ...p!, show: false }));
    }
  };

  useEffect(() => {
    if (failedNode === data.id) {
      setRunError({
        message: states?.failureReason!,
        show: false,
      });
    }
    return () => {
      setRunError(null);
    };
  }, [failedNode, data, states?.failureReason]);
  return (
    <div
      data-testid="node-wrapper"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      {runError?.show && <ErrorPopup message={runError.message} />}
      {children}
    </div>
  );
};

export default NodeWrapper;

const ErrorPopup = ({ message }: { message: string }) => {
  return (
    <div
      className="error__popup__container"
      data-testid="node-error-popup"
      style={{ top: "-50px" }}
    >
      <div className="message">{message}</div>
      <div className="error__popup__arrow"></div>
    </div>
  );
};
