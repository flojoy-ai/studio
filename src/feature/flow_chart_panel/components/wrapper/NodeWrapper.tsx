import { useFlowChartState } from "@src/hooks/useFlowChartState";
import React from "react";

const NodeWrapper = ({ children, data }) => {
  const { runningNode, failedNode } = useFlowChartState();
  return (
    <div
      style={{
        ...(runningNode === data.id && { boxShadow: "0 0 50px 15px #48abe0" }),
        ...(failedNode === data.id && {
          boxShadow: "rgb(183 0 0) 0px 0px 50px 15px",
        }),
      }}
    >
      {children}
    </div>
  );
};

export default NodeWrapper;
