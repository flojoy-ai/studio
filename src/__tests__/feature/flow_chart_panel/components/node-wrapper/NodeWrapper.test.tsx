import NodeWrapper from "@src/feature/flow_chart_panel/components/node-wrapper/NodeWrapper";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import { useSocket } from "@src/hooks/useSocket";
import { fireEvent, render } from "@testing-library/react";

jest.mock("@src/hooks/useFlowChartState");
jest.mock("@src/hooks/useSocket");

const elementData = {
  ctrls: {
    first_param: {
      functionName: "node-123",
      param: "first_param",
      value: "10",
    },
  },
  func: "Rand",
  id: "test-123",
  label: "Rand",
  type: "Default",
};

describe("NodeWrapper component", () => {
  it("Should render the component and match snapshot.", () => {
    const { container } = render(
      <NodeWrapper data={elementData}>
        <div></div>
      </NodeWrapper>
    );
    fireEvent.mouseEnter(container)
    expect(container).toMatchSnapshot();
  });
  it("Given failedNode equals to node id and failureReason, should render error popup on hover.", () => {
    const errorMessage = "Unexpected error occured!";
    (useSocket as jest.Mock).mockReturnValue({
      states: {
        failureReason: errorMessage,
      },
    });
    (useFlowChartState as jest.Mock).mockReturnValue({
        failedNode: 'test-123'
    })
    const { getByTestId } = render(
      <NodeWrapper data={elementData}>
        <div></div>
      </NodeWrapper>
    );
    const nodeWrapper = getByTestId('node-wrapper')
    fireEvent.mouseEnter(nodeWrapper)
    const errorPopupElem = getByTestId("node-error-popup");
    expect(errorPopupElem).toBeInTheDocument();
    expect(errorPopupElem).toHaveTextContent(errorMessage)
  });
});
