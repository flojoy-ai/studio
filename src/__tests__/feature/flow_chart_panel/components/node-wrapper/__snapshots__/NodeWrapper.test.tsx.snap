// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeWrapper component Should render the component and match snapshot. 1`] = `
<div>
  <div
    data-testid="node-wrapper"
  >
    <div />
  </div>
</div>
`;
